_posts/2019-05-15-the-quest-for-a-better-search.md

```diff
@@ -0,0 +1,65 @@
+
+> You are Google, you are not Google.
+
+Most of the times the “you are not Google” mantra is used to say that if you’re not at the scale of Google something (technology or architecture choices) is not worth it. The comment is not about the scale of the business, so for example in this context Amazon is not Google: Searching is one of Google’s core businesses, that’s not the case for Amazon.
+
+How does this reflect on the architecture? In a service oriented architecture:
+
+> a Service is the technical authority for a specific business capability.
+
+If searching is not a business capability, then it’s not a service in the SOA definition. Amazon for example lets you search only in product names and descriptions, it looks like a Google-like search, but it’s not. Search requests are routed to the Product Catalog service, and handled there.
+
+## Hold on, folks! What about facets?
+
+You’re right, this is not the entire story. Whenever a search is performed 2 types of results are presented:
+
+- A list of matching items/goods
+- A set of facets, usually in the left sidebar, designed to narrow the results set.
+
+![1557912046187](/img/posts/the-quest-for-a-better-search/1557912046187.png){:class="img-fluid mx-auto d-block"}
+
+From the above screenshot it’s pretty clear that those faceted results are coming from different services, not only from Product Catalog.
+
+There are even more complex scenarios when it comes to searches, think about Hotels Reservation systems. Users need to search by location and availability (if not even more criteria) and then possibly narrow the results by stars, rating, or amenities like a fitness room, or a SPA.
+
+![1557912425551](/img/posts/the-quest-for-a-better-search/1557912425551.png){:class="img-fluid mx-auto d-block"}
+
+In this scenario there are at least 2 services that participate in the search, let’s call them Hotels Catalog and Reservation. And then there are other services involved in the facets generation, for example Marketing to handle ratings, and again Hotels Catalog to handle amenities.
+
+### How can we design something like that?
+
+The simplest solution would be for all services involved in every type of supported search to:
+
+- Deliver data to a common place, for example a Sql Database or an Elasticsearch instance
+- Have the search logic backed into the aforementioned storage
+
+It works, and it’s doable. However, it violates boundaries. Is it a problem? It’s up to you to decide, it’s the typical “it depends” consultant answer.
+
+I’d say that in a small system, with a small team, with a low data change rate, and with near zero data schema evolution it’s not a problem at all. As I said: it works and brilliantly solves the problem.
+
+However it’s a huge source of coupling. As soon as the system grows, or the team gets larger, or there is higher data or data schema change rate, the hodgepodge becomes a source of pain.
```

Comment:
Is this a bigger source of coupling than UI composition? 

Personally, I'm not sure - data ownership is still inside each service and the query/search model is evolved through requirements at the system level. 

I do think this is harder to evolve form each service perspective than UI composition as we are talking about persistent state - so it's yet another storage schema to maintain.

Reply:
Data ownership is for sure inside each service, the problem with an external storage that owns schema and query (thus data and behaviors), that is what I'm describing here, is that it results in a sort of no man's land, no one will have authority to change anything. I guess it's not dissimilar from the coupling brought in by Branding in UI Composition. It's probably worse but not dissimilar.